File: stacks_and_queues/next_greater_element_i.js
var nextGreaterElement = function(nums1, nums2)
{
    let stack = [];
    let map = new Map();

    for (let i = 0; i < nums2.length; i++)
    {
        while (stack.length && nums2[i] > stack[stack.length - 1])
        {
            map.set(stack.pop(), nums2[i]);
        }

        stack.push(nums2[i]);
    }

    while (stack.length)
    {
        map.set(stack.pop(), -1);
    }

    let res = [];
    for (let i = 0; i < nums1.length; i++)
    {
        res.push(map.get(nums1[i]));
    }

    return res;
}

File: stacks_and_queues/next_greater_element_i.py
def nextGreaterElement(nums1: List[int], nums2: List[int]) -> List[int]:
    stack = []
    map = {}

    for i in range(len(nums2)):
        while stack and nums2[i] > stack[-1]:
            map[stack.pop()] = nums2[i]
        
        stack.append(nums2[i])
    
    while stack:
        map[stack.pop()] = -1
    
    res = []
    for i in range(len(nums1)):
        res.append(map[nums1[i]])
    
    return res
